### ISSUE TITLE
[DeepSeek-V4-Flash-0731] Define and validate the SGLang/Megatron train–rollout numerical parity

### ISSUE THREAD

**Issue by Suhail** [radixark/miles#2722]
We are running RLVR with `deepseek-ai/DeepSeek-V4-Flash-0731` on B200 GPUs using:

- **SGLang rollout:** TP=8, EP=8, packed FP4 experts converted to FP8 in memory, FP8 KV cache.
- **Miles/Megatron actor:** TP=8, EP=8, DP=2.
- **Trainer checkpoint:** BF16 checkpoint produced through FP4→FP8→BF16 and FP8→BF16 conversion.
- R3 MoE route replay enabled.
- Natural deterministic indexer top-k; indexer replay disabled.
- SGLang prefill-rescored behavior logprobs.
- Speculative decoding and other performance features disabled.
- Deterministic training settings.

We are having difficulty determining what level of train–rollout numerical difference is expected and safe for this configuration.

## Observed results

| Actor compute | Mean absolute logprob error | Maximum absolute error |
|---|---:|---:|
| BF16, base policy | 0.0413 | 0.0998 |
| TE blockwise FP8, base policy | 0.0666 | 0.1497 |
| BF16, trained LoRA policy | 0.0428 | 0.1318 |
| TE blockwise FP8, trained LoRA policy | 0.0660 | 0.2570 |

Additional observations:

- Almost all discrepancy occurs on the first generated token.
- EOS logprobs agree to approximately `1e-6`.
- FP8 differences fall on exact `0.125` increments.
- The FP8 mean remains stable, while the maximum changes depending on which tokens are sampled.
- TE FP8 did not improve parity over BF16 for the same checkpoint and fixed tokens.
- A strict maximum-error gate can accept or reject batches based on the sampled token set, inadvertently biasing training-data selection.

## Suspected explanation

SGLang and Megatron appear to perform different computations despite both nominally using FP8:

- SGLang uses static checkpoint FP8, FP4→FP8 expert conversion, and packed FP8 KV.
- Megatron reconstructs BF16 weights and dynamically requantizes them through Transformer Engine.
- Enabling FP8 training also enables DSV4-specific simulated FP8 behavior for KV, compressors, and indexer queries.
- Route replay aligns discrete routing decisions but does not align these continuous computations.

The observed FP8 mean appears consistent with results reported in:

- [PR #1337](https://github.com/radixark/miles/pull/1337)
- [PR #1340](https://github.com/radixark/miles/pull/1340)

We also incorporated the relevant precision fixes from:

- [PR #1505](https://github.com/radixark/miles/pull/1505)
- [PR #1584](https://github.com/radixark/miles/pull/1584)

## Questions

1. What precision pairing is recommended for the 0731 checkpoint?

   - FP8 SGLang rollout with BF16 Megatron training?
   - FP8 rollout with TE blockwise FP8 training?
   - A separately materialized, unified FP8 rollout checkpoint?

2. Which measurements and thresholds should determine whether training is safe to start?

   - Mean absolute logprob difference?
   - p99 absolute difference?
   - Approximate KL?
   - Pearson correlation?
   - A hard per-token maximum?

3. Could Miles provide a fixed-token parity example with expected numerical ranges on B200?

4. What exact conversion command, Transformer Engine precision configuration, environment variables, and SGLang flags are recommended for 0731?

5. Should parity reporting separately identify differences caused by:

   - static-weight FP8,
   - TE GEMM requantization,
   - DSV4 QAT for KV/compressors/indexer queries,
   - routing differences?

6. Could the official recipe expose fixed-token mean, p99, and KL trend monitoring so that changing sampled tokens do not make a maximum-error gate unstable?

## Current workaround

Our current conservative approach is:

- Keep BF16 actor compute.
- Use clipped GRPO with importance-ratio and KL telemetry.
- Monitor parity on a fixed token set.
- Treat FP8 training as unvalidated for this pinned 0731 stack.

**Comment by Suhail** [comment]
May someone's agent find this useful:

## DeepSeek-V4-Flash-0731 + Miles/SGLang: what got train/inference logprob mismatch near zero

We eventually reached the following result on a short RLVR arithmetic canary:

- Miles actor logprob: approximately `-0.02190`
- SGLang rollout logprob: approximately `-0.01668`
- Mean absolute logprob difference: `0.0052195`
- K3 KL: `0.00002718`
- MoE route replay: disabled
- Indexer replay: disabled
- Evaluation reward: `122–124/128`

This was not one magic fix. We had to align checkpoint interpretation, numerical kernels, token/logprob semantics, adapter synchronization, and policy versions.

### Important qualification

This is near-zero parity on a short, saturated two-token answer after the model had overfit the toy task. It is not proof that Miles and SGLang are universally bit-exact.

Earlier, less-confident policies still showed approximately `0.05–0.13` mean error, and a longer diagnostic response had approximately `0.0807` mean / `0.316` maximum error.

---

## Working configuration

### Model and topology

- Model: `deepseek-ai/DeepSeek-V4-Flash-0731`
- Revision: `7872f01b1d1fe23eabc4c98b48bffcef5a386062`
- Hardware: NVIDIA B200
- Trainer: Miles/Megatron, two nodes
- Trainer topology: `TP=8, EP=8, PP=1, DP=2`
- Rollout: two independent SGLang nodes
- Each rollout node: `TP=8, EP=8`
- LoRA: rank 16 on `wq_a`

### SGLang rollout

- Original untouched 0731 checkpoint
- Packed FP4 experts converted to FP8 in memory
- Existing FP8 tensors retained
- FP8 E4M3 KV cache
- Speculative decoding disabled
- CUDA graphs disabled
- Overlap scheduling disabled
- Radix cache disabled

### Miles actor

- Policy-only BF16 checkpoint decoded from the original checkpoint
- BF16 parameters and trusted baseline compute
- BF16 LoRA
- BF16 DSA indexer projection
- FP32 softmax and important reductions
- Transformer Engine blockwise FP8 tested separately, but BF16 retained as the default

---

## 1. Interpret the 0731 checkpoint correctly

The original checkpoint is mixed-format:

- Routed experts are packed FP4.
- Other policy tensors are FP8, BF16, or FP32.
- MTP/DSpark tensors are also present.

Miles could not directly train the original packed FP4 tensors. We therefore:

1. Used SGLang’s own lossless FP4-to-FP8 decoding path.
2. Dequantized those FP8 values into BF16 for Miles.
3. Dequantized the checkpoint’s other FP8 policy tensors into BF16.
4. Preserved native FP32 tensors as FP32.
5. Excluded MTP/DSpark tensors because speculation was disabled.
6. Converted the resulting policy into Megatron `torch_dist` format at TP8/EP8/PP1.
7. Verified that every source tensor was converted, copied, or explicitly excluded.

SGLang continued to use the original checkpoint with:

```bash
SGLANG_DSV4_FP4_DEQUANT=1
```

Do **not** set:

```bash
SGLANG_DSV4_FP4_EXPERTS=0
```

against the original packed checkpoint. That causes the packed FP4 storage to be interpreted as ordinary FP8 tensors.

We also disabled the optimized `wo_a` FP8 GEMM while establishing the baseline:

```bash
SGLANG_OPT_FP8_WO_A_GEMM=0
```

---

## 2. Establish frozen repeatability before testing parity

Before comparing Miles to SGLang, we made each system repeat itself.

For each rollout server we required:

- Identical prompt and token IDs
- Three repeated executions
- Identical generated token IDs
- Byte-identical float32 logprobs
- No restart, OOM, CUDA, or NCCL error

For deterministic SGLang validation, we also needed:

- DSV4 added to SGLang’s deterministic-attention allowlist
- BF16 hash-router logits converted to FP32 before the fused hash-topk kernel
- Maximum admitted context explicitly prewarmed after every server restart

The long-context DSV4 path had a one-time shape transition: a 4096-token request could change both its first result and a previously stable 2304-token result. After a 4096-token prewarm, both rollout servers became byte-repeatable at 2304 and 4096 tokens.

Without frozen repeatability, actor/rollout comparison results are ambiguous.

---

## 3. Fix or avoid FP8 scale-format mismatches

A material early improvement came from respecting the checkpoint’s UE8M0 scale format in SGLang’s Triton MoE path.

The original route dynamically quantized activations without consistently propagating that the model expected UE8M0, power-of-two scales. We propagated this setting through:

- The FP8 quantization configuration
- FC1 activation quantization
- FC2 activation quantization
- The Triton MoE runner

This materially improved parity.

This is backend-specific. Anyone using the Triton MoE path with the 0731 checkpoint should verify that the dynamic activation-scale format matches the checkpoint’s `scale_fmt`.

---

## 4. Handle the DSA indexer separately

Longer prompts exposed another independent issue:

- The fused indexer top-k drifted once C4 had more than approximately 512 candidates.
- Switching only the final top-k operation to `torch.topk` was insufficient.
- The paged-MQA scoring kernel could already produce different scores before top-k.

Our long-context correctness configuration used:

```bash
SGLANG_TOPK_TRANSFORM_512_TORCH=1
SGLANG_FP8_PAGED_MQA_LOGITS_TORCH=1
```

We also fixed a shape mismatch where `seq_lens` arrived as `[B, 1]` while the PyTorch fallback expected `[B]`.

The final near-zero toy prompt contained only 28 real tokens. At that length, every available compressed key was selected, so indexer selection was effectively not a meaningful branch. The short-prompt result should not be treated as proof of long-context DSA parity.

---

## 5. Compare exactly the same probability

We explicitly ruled out logprob extraction mistakes by aligning:

- Exact prompt token IDs
- Exact response token IDs
- Chat template and `</think>` boundary
- Response-token shift convention
- Loss mask
- Prompt versus response positions
- Padding treatment
- Selected-token logprob semantics
- Temperature and sampling metadata
- Policy version

We compared only the response tokens selected by the loss mask.

We also checked SGLang incremental decode against fixed-token prefill rescoring. Every first response-token logprob matched exactly, and the largest remaining difference was approximately `1.7e-6` on a near-zero EOS logprob. Incremental versus prefill execution was therefore not the main actor/rollout mismatch.

---

## 6. Use routing replay as a diagnostic, not the final solution

Initially, natural routing produced very large disagreement:

| Mode | Mean absolute error | Maximum error |
|---|---:|---:|
| Natural MoE routing | `0.369` | `3.55` |
| Expert-ID replay | `0.139` | `0.537` |

This showed that routing mattered, but also that expert-ID replay was incomplete.

SGLang’s routed MoE decision contains:

- Six expert IDs
- Six continuous FP32 routing weights
- Six-slot ordering

Miles’ original replay path forced the expert IDs but recomputed the continuous weights, normalization, scaling, and dense representation. Identical IDs therefore did not imply identical MoE output.

We captured complete `[tokens, layers, 6]` traces containing IDs, FP32 weights, and slot order for debugging. However, exact replay is expensive and not scalable, so we did not make it the production solution.

After the policy and surrounding numerical paths were stabilized, we removed both:

- MoE route replay
- Indexer replay

The replay-free run then produced the final `0.00522` mean difference.

This does **not** prove that routing differences disappeared globally. On the saturated toy answer, both implementations assigned nearly all probability to the same token, making internal routing differences less visible at the final logprob.

---

## 7. Make adapter transfer exact and version-safe

A surprisingly easy way to create “numerical mismatch” is to compare two different policy versions.

Our LoRA exporter:

1. Gathered all TP shards.
2. Exported exactly 86 BF16 tensors for 43 `wq_a` adapters.
3. Validated tensor names, shapes, dtypes, and finiteness.
4. Compared corresponding adapter tensors across both DP replicas.
5. Sent expected tensor checksums to SGLang.
6. Hot-loaded both rollout servers.
7. Rejected samples with missing or mixed policy versions.

Both rollout servers had to acknowledge the new version before their trajectories were accepted.

We also checked that every saved checkpoint strictly reloaded and rejected:

- Missing tensors
- Extra tensors
- Incorrect shapes
- Incorrect dtypes
- Nonfinite tensors
- An entirely zero adapter
- Zero matched tensors

---

## 8. Use realistic parity gates

Our original gate was too strict and rejected roughly 7 of 11 batches. This:

- Wasted cold trainer startups
- Reduced training throughput
- Biased training toward batches with unusually easy numerical parity

We switched to a fail-closed but practical gate based on:

- Mean absolute logprob difference
- K3 KL
- All values finite
- Exact token alignment
- Exact policy-version alignment

Our stop thresholds were:

```text
mean absolute logprob difference <= 0.30
K3 KL <= 0.10
```

Those are safety limits, not target values. The final BF16 result was much better:

```text
mean absolute difference = 0.0052195
K3 KL                  = 0.00002718
```

Using mean error and KL was much more useful than demanding bit-identical logprobs or gating primarily on the single worst token.

---

## 9. Train until the canary became confident

This was an important part of the final near-zero result.

We ran 19 real BF16 GRPO updates. Fixed-seed 128-sample evaluations progressed approximately as follows:

```text
33/128
...
119/128
119/128
118/128
123/128
```

With natural routing and no replay, later evaluations reached:

```text
122/128
123/128
124/128
```

At that point, both Miles and SGLang assigned nearly all probability to the correct response. Their logprobs were both near zero:

```text
Miles:  -0.02190
SGLang: -0.01668
```

That confidence saturation contributed materially to the final small absolute difference.

---

## Things that did not fix the mismatch

These negative results may save others time:

| Experiment | Result |
|---|---|
| Round KV inputs through BF16 before FP8 cache quantization | Worse: approximately `0.190` mean / `2.786` max |
| Globally change actor compute from TE FP8 to BF16 | Did not fix earlier difficult batches; BF16 was sometimes worse |
| Force router output to FP32 alone | Worse |
| Change expert FC2 accumulation to K=32 | Rollout logprobs remained bit-identical; not the cause |
| Move route-weight multiplication outside the expert kernel | Substantially worse |
| Replay only expert IDs | Helpful diagnostically, but incomplete |
| Blame SGLang decode versus prefill | Ruled out; they matched within approximately `1.7e-6` |
| Disable or replace the KV cache | Not the leading cause |
| Enable speculative decoding | Never enabled during correctness work |

Transformer Engine blockwise FP8 was also not universally near-zero:

```text
Frozen FP8 canary:
  mean abs difference = 0.01915
  KL                  = 0.000362

Two real FP8 updates:
  mean abs difference = 0.1289 and 0.1106
  KL                  = 0.02284 and 0.02168
```

Both FP8 steps were finite and completed successfully, but BF16 remains our trusted actor baseline.

---

## Recommended debugging order for other implementations

1. Pin the exact model, Miles, SGLang, image, tokenizer, and topology revisions.
2. Prove that every checkpoint tensor is interpreted correctly.
3. Make each rollout server independently repeatable.
4. Compare the same fixed response tokens rather than separately sampled outputs.
5. Verify token shifting, masking, and selected-logprob semantics.
6. Verify adapter checksums and policy versions on every rollout server.
7. Start with conservative BF16 actor compute.
8. Disable speculative decoding, CUDA graphs, overlap scheduling, and other optimization paths.
9. Test MoE routing and DSA indexer behavior independently.
10. Use replay only to isolate a subsystem, not as the permanent solution.
11. Gate training on finite mean error and KL rather than exact equality.
12. Test both a saturated short prompt and a low-confidence or long-context prompt.
13. Treat matched FP8 training and long-context DSA parity as separate graduation gates.

The most important lesson is that a small final logprob gap can come from both improved numerical alignment **and** policy saturation. Always report the prompt length, response length, policy confidence, per-token errors, and KL alongside the aggregate mismatch.
